agent/src/main/java/com/appland/appmap/process/hooks/Proxy.java

```diff
@@ -0,0 +1,155 @@
+package com.appland.appmap.process.hooks;
+
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+import org.tinylog.TaggedLogger;
+
+import com.appland.appmap.config.AppMapConfig;
+import com.appland.appmap.output.v1.Event;
+import com.appland.appmap.record.EventTemplateRegistry;
+import com.appland.appmap.transform.ClassFileTransformer;
+import com.appland.appmap.transform.annotations.ArgumentArray;
+import com.appland.appmap.transform.annotations.Hook;
+import com.appland.appmap.transform.annotations.HookClass;
+import com.appland.appmap.transform.annotations.HookSite;
+import com.appland.appmap.transform.annotations.MethodEvent;
+import com.appland.appmap.util.AppMapClassPool;
+import com.appland.appmap.util.ClassUtil;
+import com.appland.appmap.util.FullyQualifiedName;
+
+import javassist.CtBehavior;
+import javassist.NotFoundException;
+
+public class Proxy {
+  private static final TaggedLogger logger = AppMapConfig.getLogger(null);
+  private static final Map<Method, HookSite[]> allHookSites = new HashMap<>();
```

Comment:
Should this be an `Map<Method, EnumMap<MethodEvent, HookSite>>` instead? The code is unnecessarily opaque with all those `toIndex()` and integer indices.

Or even a specialized class holding the `HookSite` as members, seeing as their catalog is rather static, and having those three `invoke`, `return`, `exc` methods that would build the event and invoke the handler.